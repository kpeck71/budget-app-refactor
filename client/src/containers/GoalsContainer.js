import React, { Component } from 'react';
import axios from 'axios'
import Goal from '../components/Goal';
import NewGoalForm from '../components/NewGoalForm';
import EditGoalForm from '../components/EditGoalForm';
import { connect } from 'react-redux';
import * as actions from '../actions/actions';

class GoalsContainer extends Component {
  constructor(props) {
    super(props)
    this.state = {
      goals: [],
      editingGoalId: null
    }
    this.addNewGoal = this.addNewGoal.bind(this)
    this.removeGoal = this.removeGoal.bind(this)
    this.editingGoal = this.editingGoal.bind(this)
    this.editGoal = this.editGoal.bind(this)
  }

  componentDidMount() {
    axios.get('http://localhost:3001/api/v1/goals.json')
    .then(response => {
      console.log(response)
      this.setState({
        goals: response.data
      })
    })
    .catch(error => console.log(error))
  }

  addNewGoal(title, amount) {
     axios.post( '/api/v1/goals', { goal: {title, amount} })
     .then(response => {
         console.log(response)
         const goals = [ ...this.state.goals, response.data ]
         this.setState({goals})
     })
     .catch(error => {
         console.log(error)
     })
   }

   removeGoal(id) {
      axios.delete( '/api/v1/goals/' + id )
      .then(response => {
          const goals = this.state.goals.filter(
              goal => goal.id !== id
          )
          this.setState({goals})
      })
      .catch(error => console.log(error))
    }

    editingGoal(id) {
      this.setState({
        editingGoalId: id
      })
    }

    editGoal(id, title, amount) {
      axios.put( '/api/v1/goals/' + id, {
          goal: {
              title,
              amount
          }
      })
      .then(response => {
          console.log(response);
          const goals = this.state.goals;
          goals[id-1] = {id, title, amount}
          this.setState(() => ({
              goals,
              editingGoalId: null
          }))
      })
      .catch(error => console.log(error));
    }

  render() {
    return (
      <div className="goals-container container-fluid">
        <div className="row">
          <h1>Goals</h1>
        {this.state.goals.map( goal => {
          if (this.state.editingGoalId === goal.id ) {
            return (<EditGoalForm
                      goal={goal}
                      key={goal.id}
                      editGoal={this.editGoal}
                    />)
          } else {
          return (<Goal
                    goal={goal}
                    key={goal.id}
                    onRemoveGoal={this.removeGoal}
                    editingGoal={this.editingGoal}
                    /> )
          }
        })}
        </div>
        <NewGoalForm onNewGoal={this.addNewGoal} />
      </div>
      )
    }
  }
const mapStateToProps = state => ({
  goals: state.goals.goals,
  loading: state.goals.loading,
  error: state.goals.error

})

export default connect (mapStateToProps, actions)(GoalsContainer);
