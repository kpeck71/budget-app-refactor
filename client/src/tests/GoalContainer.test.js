import { shallow } from "enzyme";
import React from "react";
import ReactDOM from "react-dom";
import GoalsContainer from "../containers/GoalsContainer"
import './enzyme_setup'

describe("GoalsContainer", () => {
  it("renders without crashing", () => {
    const goals = []
    const div = document.createElement("div");
    ReactDOM.render(<GoalsContainer goals={goals} />, div)
  });

  it("should render even with no goals", () => {
    const goals = []
    const wrapper = shallow(
      <GoalsContainer goals={goals} />
    );
    expect(wrapper.find(".goals-container").exists()).toBe(true);
  });
})

describe("when given `goals`", () => {
  const goals = [
    { id: 1, title: "Trip to Paris", amount: 2000 },
    { id: 2, title: "New bike", amount: 1000}
  ]

  it("should render correct number of goals", () => {
    const wrapper = shallow(
      <GoalsContainer goals={goals} />
    );
    expect(wrapper.find("title")).toMatchSnapshot();
    // expect(wrapper.find(".single-goal").length).toEqual(2)
  });
})
