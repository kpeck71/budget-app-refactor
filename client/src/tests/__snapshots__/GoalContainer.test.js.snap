// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when given \`goals\` should render correct number of goals 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): ShallowWrapper {
    Symbol(enzyme.__root__): [Circular],
    Symbol(enzyme.__unrendered__): <GoalsContainer
      goals={
        Array [
          Object {
            "amount": 2000,
            "id": 1,
            "title": "Trip to Paris",
          },
          Object {
            "amount": 1000,
            "id": 2,
            "title": "New bike",
          },
        ]
      }
    />,
    Symbol(enzyme.__renderer__): Object {
      "batchedUpdates": [Function],
      "getNode": [Function],
      "render": [Function],
      "simulateError": [Function],
      "simulateEvent": [Function],
      "unmount": [Function],
    },
    Symbol(enzyme.__node__): Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="row"
          >
            <h1>
              Goals
            </h1>
          </div>,
          <NewGoalForm
            onNewGoal={[Function]}
          />,
        ],
        "className": "goals-container container-fluid",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h1>
                Goals
              </h1>,
              Array [],
            ],
            "className": "row",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Goals",
              },
              "ref": null,
              "rendered": "Goals",
              "type": "h1",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "onNewGoal": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
    Symbol(enzyme.__nodes__): Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="row"
            >
              <h1>
                Goals
              </h1>
            </div>,
            <NewGoalForm
              onNewGoal={[Function]}
            />,
          ],
          "className": "goals-container container-fluid",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <h1>
                  Goals
                </h1>,
                Array [],
              ],
              "className": "row",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Goals",
                },
                "ref": null,
                "rendered": "Goals",
                "type": "h1",
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "onNewGoal": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    Symbol(enzyme.__options__): Object {
      "adapter": ReactSixteenAdapter {
        "options": Object {
          "enableComponentDidUpdateOnSetState": true,
          "lifecycles": Object {
            "componentDidUpdate": Object {
              "onSetState": true,
            },
            "getDerivedStateFromProps": true,
            "getSnapshotBeforeUpdate": true,
            "setState": Object {
              "skipsComponentDidUpdateOnNullish": true,
            },
          },
        },
      },
    },
  },
  Symbol(enzyme.__unrendered__): null,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): undefined,
  Symbol(enzyme.__nodes__): Array [],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__rootNodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="row"
          >
            <h1>
              Goals
            </h1>
          </div>,
          <NewGoalForm
            onNewGoal={[Function]}
          />,
        ],
        "className": "goals-container container-fluid",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h1>
                Goals
              </h1>,
              Array [],
            ],
            "className": "row",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Goals",
              },
              "ref": null,
              "rendered": "Goals",
              "type": "h1",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "onNewGoal": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
}
`;
